doc/features/speculative-executions/README.md

```diff
@@ -0,0 +1,214 @@
+  |<---------------|                     |
+  |                | cancel              |
+  |                |-------------------->|
+```
+
+Note that "cancelling" in this context simply means marking the operation to discard the response when it later arrives.
+
+Speculative executions are disabled by default. The following sections cover the practical details and how to
+enable them.
+
+## Query idempotence
+
+One important aspect to consider is whether queries are idempotent, (that is, whether they can be applied multiple
+times without changing the result beyond the initial application). If a query is not idempotent, the driver never
+schedules speculative executions for it, because there is no way to guarantee that only one node will apply the
+mutation.
+
+Examples of queries that are not idempotent are:
+
+- counter operations
+- prepending or appending to a list column
+- using non-idempotent CQL functions, like `now()` or `uuid()`
+
+In the driver, this is determined by [`isIdempotent` flag in the `QueryOptions`][query-options-api]. Because the driver 
+does not parse query strings, in most cases it has no information about what the query actually does. Therefore, for
+ all other types of statements, it defaults to `false`. You must set it manually with one of the mechanisms described
+  below.
+
+You can override the value for each execution:
+
+```js
+const query = 'SELECT * FROM users WHERE key = ?';
+client.execute(query, [ 'usr1' ], { prepare: true, isIdempotent: true }); 
+
+```
+
+Additionally, if you know for a fact that your application does not use any of the non-idempotent CQL queries listed 
+above, you can change the default cluster-wide:
+
+```js
+// Make all statements idempotent by default:
+const client = new Client({
+  contactPoints,
+  queryOptions: { 
+    isIdempotent: true
+  }
+});
+```
+
+## Enabling speculative execution 
+
+Speculative executions are controlled by an instance of `SpeculativeExecutionPolicy` provided when initializing the
+Client. This policy defines the threshold after which a new speculative execution is triggered.
+
+The driver provides a `ConstantSpeculativeExecutionPolicy` that schedules a given number of speculative executions,
+separated by a fixed delay, the policy is exported under the `<root>.policies.speculativeExecution` module.
+
+This simple policy uses a constant threshold:
+
+```js
+const client = new Client({
+  contactPoints,
+  policies: {
+    speculativeExecution: new ConstantSpeculativeExecutionPolicy(
+      200, // delay before a new execution is launched
+      2) // maximum amount of additional executions
+  }
+});
+```
+
+Given the configuration above, an idempotent query would be handled this way:
+
+- start the initial execution at t0
+- if no response has been received at t0 + 200 milliseconds, start a speculative execution on another node
+- if no response has been received at t0 + 400 milliseconds, start another speculative execution on a third node
```

Comment:
Similar to comment above, if the configuration is `2`, I would only expect there to be one execution beyond the initial one.

Reply:
I just noticed that the java documentation is wrong from this perspective as well.  Will fix it there.

Reply:
The second parameter of `ConstantSpeculativeExecutionPolicy` is the amount of _speculative executions_, I think the legend is correct.

The `SpeculativeExecutionPolicy` is first used to get a plan after the first execution.

Reply:
Ack I'm wrong about that.  I get confused by that constantly.  In that case, i like `additional` as it is more explicit to mean beyond the first.

Reply:
It would be good to use that in the `ConstantSpeculativeExecutionPolicy` jsdoc too to make it more clear that it means beyond the first.

Reply:
will do